8311821: Simplify ParallelGCThreadsConstraintFunc after CMS removal

kspeeyu opened the pull request:
Code oddities reported by SonarCloud after CMS removal
Fix: moving ParallelGCThreadsConstraintFuncParallel to shared code as ParallelGCThreadsConstraintFunc

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311821](https://bugs.openjdk.org/browse/JDK-8311821): Simplify ParallelGCThreadsConstraintFunc after CMS removal (**Enhancement** - P4)


### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14965/head:pull/14965` \
`$ git checkout pull/14965`

Update a local copy of the PR: \
`$ git checkout pull/14965` \
`$ git pull https://git.openjdk.org/jdk.git pull/14965/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14965`

View PR using the GUI difftool: \
`$ git pr show -t 14965`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14965.diff">https://git.openjdk.org/jdk/pull/14965.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14965#issuecomment-1644785981)

kimbarrett commented on src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp:
```diff
@@ -53,18 +53,19 @@
-  status = ParallelGCThreadsConstraintFuncParallel(value, verbose);
-  if (status != JVMFlag::SUCCESS) {
-    return status;
+  if (UseParallelGC && (value > (uint)max_jint)) {
```
Throughout this change: s/max_jint/INT_MAX/

kimbarrett commented on src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp:
```diff
@@ -53,18 +53,19 @@
 // the flag has been set by the user and so should be checked.
 
-// As ParallelGCThreads differs among GC modes, we need constraint function.
+// (earlier ParallelGCThreadConstraintsFuncParallel) replaced after CMS is removed
```
I don't think we need this comment.

kimbarrett commented on src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp:
```diff
@@ -53,18 +53,19 @@
+                        "ParallelGCThreads (" UINT32_FORMAT ") must be "
+                        "less than or equal to " UINT32_FORMAT " for Parallel GC\n",
+                        value, max_jint);
+    return JVMFlag::VIOLATES_CONSTRAINT;
```
There is no reason to use UINT32_FORMAT here, and indeed it just clutters things and makes it harder to read
and understand (and could even be wrong on some theoretical platform with ints larger than 32bits).
Just using "%u" for the first, and "%d" for the second is simple and correct.

kimbarrett reviewed (changes requested):
Could we eliminate the constraint function and instead specify a range for the option?
Using `range(0, INT_MAX)` would reduce the upper bound for collectors other than
ParallelGC, but I doubt anyone would notice :)

Or keep the constraint function if we can also move the check here that the value must be
greater than zero for most (i.e. not SerialGC) collectors.  (I think that's in ergo and might
not be movable because of other ergo calculations based on it that might not work for
a value of 0.)

shipilev commented:
> Could we eliminate the constraint function and instead specify a range for the option? Using `range(0, INT_MAX)` would reduce the upper bound for collectors other than ParallelGC, but I doubt anyone would notice :)

I agree! Let's just put the `range(0, INT_MAX)` here: https://github.com/openjdk/jdk/blob/8cd43bff3cd18d6e83cbf07b78a809ad002993c5/src/hotspot/share/gc/shared/gc_globals.hpp#L136

shipilev commented:
I don't think we need `ParallelGCThreadsConstraintFunc` at all now, seeing how we removed its only (?) use.

kspeeyu commented:
> I don't think we need `ParallelGCThreadsConstraintFunc` at all now, seeing how we removed its only (?) use.

okay, so we remove 'ParallelGCThreadsConstraintFunc' as we replaced it with specifying a range instead? got it.

kimbarrett commented on src/hotspot/share/gc/shared/gc_globals.hpp:
```diff
@@ -133,7 +133,7 @@
   product(uint, ParallelGCThreads, 0,                                       \
           "Number of parallel threads parallel gc will use")                \
-          constraint(ParallelGCThreadsConstraintFunc,AfterErgo)             \
+          range(0, INT_MAX)                                                 \
```
A comment to explain the range restriction seems useful.

shipilev replied:
+1. Something like:

```
  /* notice: the max range value here is INT_MAX not UINT_MAX */            \
  /* to protect from overflows                                */            \
  product(uint, ParallelGCThreads, 0,                                       \
          "Number of parallel threads parallel gc will use")                \
          range(0, INT_MAX)                                                 \
```

kimbarrett reviewed (approved):
Other than the suggestion for an additional comment, this looks good.

shipilev reviewed (approved):
Looks good, with minor nit about overflow comment.

shipilev commented:
Ah, see, there is a failing test:
https://github.com/kspeeyu/jdk/actions/runs/5651211226#user-content-gc_arguments_testparallelgcthreads

```
----------System.err:(22/1299)----------
 stdout: [];
 stderr: [uint ParallelGCThreads=4294967295 is outside the allowed range [ 0 ... 2147483647 ]
Improperly specified VM option 'ParallelGCThreads=4294967295'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
]
 exitValue = 1
```

We need to adjust that test for new boundary now, I think.

shipilev commented on test/hotspot/jtreg/gc/arguments/TestParallelGCThreads.java:
```diff
@@ -113,14 +113,14 @@ public static void testFlags() throws Exception {
 
-    // 4294967295 == (unsigned int) -1
-    // So setting ParallelGCThreads=4294967295 should give back 4294967295
+    // 2147483647 == (signed int) -1
```
This is not technically correct: `(signed int) -1` is `-1`.
Rewrite the whole comment to: `// Test the max value for ParallelGCThreads`.

shipilev reviewed (approved):
This looks okay. @kimbarrett, you good with modifying the test a little?

shipilev commented:
Or maybe @tschatzl ?

kimbarrett reviewed (approved):
Still looks good.

kspeeyu commented:
/integrate

shipilev commented:
/sponsor
